lib/service/awsoidc.go

```diff
@@ -0,0 +1,263 @@
+			if err != nil {
+				process.log.Warningf("Failed to update deploy service agents: %v", err)
+				continue
+			}
+
+			ownershipTags := map[string]string{
+				types.ClusterLabel:     clusterName,
+				types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+				types.IntegrationLabel: integrationName,
+			}
+
+			// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+			// If the lease cannot be acquired, the update is already being handled by another instance.
+			semLock, err := authClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.SemaphoreKindConnection,
+				SemaphoreName: fmt.Sprintf("update_deploy_service_agents_%s_%s", region, integrationName),
+				MaxLeases:     1,
+				Expires:       process.Clock.Now().Add(updateDeployAgentsInterval),
+				Holder:        "update_deploy_service_agents",
+			})
+
+			if err != nil {
+				if strings.Contains(err.Error(), teleport.MaxLeases) {
+					process.log.Debug("Deploy service agent update is already being processed")
+					continue
+				}
+				return trace.Wrap(err)
+			}
+
+			if err := awsoidc.UpdateDeployServiceAgents(ctx, deployServiceClient, clusterNameConfig.GetClusterName(), teleportVersion, ownershipTags); err != nil {
+				process.log.Warningf("Failed to update deploy service agents: %v", err)
+
+				// Release the semaphore lease on failure so that another instance may attempt the update
+				if err := authClient.CancelSemaphoreLease(ctx, *semLock); err != nil {
+					process.log.WithError(err).Error("Failed to cancel semaphore lease")
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// getStableTeleportVersion returns the current stable version of teleport
+func getStableTeleportVersion(ctx context.Context) (string, error) {
+	var versionEndpoint string
+	var err error
+	if automaticupgrades.GetChannel() != "" {
+		versionEndpoint, err = url.JoinPath(automaticupgrades.GetChannel(), "version")
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+	}
+
+	stableVersion, err := automaticupgrades.Version(ctx, versionEndpoint)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	// cloudStableVersion has vX.Y.Z format, however the container image tag does not include the `v`.
+	return strings.TrimPrefix(stableVersion, "v"), nil
+}
+
+// withinUpgradeWindow returns true if the current time is within the configured
+// upgrade window.
+func withinUpgradeWindow(cmc types.ClusterMaintenanceConfig, clock clockwork.Clock) bool {
+	upgradeWindow, ok := cmc.GetAgentUpgradeWindow()
+	if !ok {
+		return false
+	}
+
+	now := clock.Now()
+	if len(upgradeWindow.Weekdays) == 0 {
+		if int(upgradeWindow.UTCStartHour) == now.Hour() {
+			return true
+		}
+	}
+
+	weekday := now.Weekday().String()
+	for _, upgradeWeekday := range upgradeWindow.Weekdays {
+		if weekday == upgradeWeekday {
+			if int(upgradeWindow.UTCStartHour) == now.Hour() {
+				return true
+			}
+		}
+	}
+	return false
+}
```

Comment:
I think this method belongs to something more generic :thinking: 
Maybe `lib/automaticupgrades` ?